Sources/NIOCertificateHotReloading/NIOCertificateHotReloading.swift

```diff
@@ -0,0 +1,301 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2025 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftASN1
+import NIOConcurrencyHelpers
+import NIOSSL
+import ServiceLifecycle
+import X509
+
+import struct NIOCore.TimeAmount
+
+#if canImport(FoundationEssentials)
+import FoundationEssentials
+#else
+import Foundation
+#endif
+
+/// A protocol that defines a certificate reloader.
+///
+/// A certificate reloader is a service that can provide you with updated versions of a certificate and private key pair, in
+/// the form of a `NIOSSLContextConfigurationOverride`, which will be used when performing a TLS handshake in NIO.
+/// Each implementation can choose how to observe for changes, but they all require an ``sslContextConfigurationOverride``
+/// to be exposed.
+@available(macOS 11.0, iOS 14, tvOS 14, watchOS 7, *)
+public protocol CertificateReloader: Sendable {
+    /// A `NIOSSLContextConfigurationOverride` that will be used as part of the NIO application's TLS configuration.
+    /// Its certificate and private key will be kept up-to-date via whatever mechanism the specific ``CertificateReloader``
+    /// implementation provides.
+    var sslContextConfigurationOverride: NIOSSLContextConfigurationOverride { get async }
```

Comment:
I have been thinking about this some more. There is two ways to go with this here. Keep it close to NIO then it should probably not be async but rather returning an ELF. Or this is a protocol that we see as a bridge to the Swift Concurrency ecosystem then we should make it async but I also think we should be returning a type here that is based on swift-certificates and map to NIOSSL types under the hood.

I think in general this should become a method though and not a computed property.

Would be interested to hear @Lukasa thoughts here

Reply:
Isn't the intention for this to effectively be a current value? The reloader implementation does any work it actually needs be it async or otherwise.

Reply:
In theory someone may want to avoid having a long running `run` for some reason?
Or another possible scenario would be if you were to yield new pairs into an async stream continuation from the `run` method, and iterate the stream in this getter instead of updating some [protected] state (like a LockedValueBox) containing the new values.

In any case, I do think that this is all avoidable and you can handle the asynchrony from elsewhere (like a `run` method). But not sure if Franz had something else in mind. I suppose having it `async` gives implementers some more freedom.

Reply:
I'd prefer if it's not async. Otherwise calling it from synchronous code requires more complicated state management, etc.

Reply:
I agree with George's assessment - my understanding was that this represents the "current" value. If so, I think it should be non-throwing, and sync.

Reply:
It _is_ meant to represent the current state yes. After this discussion and some more thought I'm also thinking I should revert this to be sync.